File: binary_search_tree/binary_search_tree.go
package main

import "fmt"

type BST struct {
	value       int
	left, right *BST
}

func NewBST(value int) *BST {
	return &BST{value: value}
}

func (bst *BST) Insert(value int) *BST {
	current := bst
	for {
		if value < current.value {
			if current.left == nil {
				current.left = NewBST(value)
				break
			} else {
				current = current.left
			}
		} else {
			if current.right == nil {
				current.right = NewBST(value)
				break
			} else {
				current = current.right
			}
		}
	}
	return bst
}

func (bst *BST) Contains(value int) bool {
	current := bst
	for current != nil {
		if value < current.value {
			current = current.left
		} else if value > current.value {
			current = current.right
		} else {
			return true
		}
	}
	return false
}

func (bst *BST) Remove(value int, parent *BST) *BST {
	current := bst
	for current != nil {
		if value < current.value {
			parent = current
			current = current.left
		} else if value > current.value {
			parent = current
			current = current.right
		} else {
			// Node with two children
			if current.left != nil && current.right != nil {
				current.value = current.right.GetMinValue()
				current.right.Remove(current.value, current)
			} else if parent == nil { // root node
				if current.left != nil {
					current.value = current.left.value
					current.right = current.left.right
					current.left = current.left.left
				} else if current.right != nil {
					current.value = current.right.value
					current.left = current.right.left
					current.right = current.right.right
				} else {
					// Single-node tree, do nothing
				}
			} else if parent.left == current {
				if current.left != nil {
					parent.left = current.left
				} else {
					parent.left = current.right
				}
			} else if parent.right == current {
				if current.left != nil {
					parent.right = current.left
				} else {
					parent.right = current.right
				}
			}
			break
		}
	}
	return bst
}

func (bst *BST) GetMinValue() int {
	current := bst
	for current.left != nil {
		current = current.left
	}
	return current.value
}

func main() {
	bst := NewBST(10)
	bst.Insert(5).Insert(4).Insert(6).Insert(15).Insert(12).Insert(17)

	bst.Remove(10, nil)
	fmt.Println(bst.Contains(10)) // false
}

File: binary_search_tree/binary_search_tree.java
public class BST {
    public int value;
    public BST left;
    public BST right;

    public BST(int value) {
        this.value = value;
    }

    public BST insert(int value) {
        BST current = this;
        while (true) {
            if (value < current.value) {
                if (current.left == null) {
                    current.left = new BST(value);
                    break;
                } else {
                    current = current.left;
                }
            } else {
                if (current.right == null) {
                    current.right = new BST(value);
                    break;
                } else {
                    current = current.right;
                }
            }
        }
        return this;
    }

    public boolean contains(int value) {
        BST current = this;
        while (current != null) {
            if (value < current.value) {
                current = current.left;
            } else if (value > current.value) {
                current = current.right;
            } else {
                return true;
            }
        }
        return false;
    }

    public BST remove(int value, BST parent) {
        BST current = this;
        while (current != null) {
            if (value < current.value) {
                parent = current;
                current = current.left;
            } else if (value > current.value) {
                parent = current;
                current = current.right;
            } else {
                if (current.left != null && current.right != null) {
                    current.value = current.right.getMinValue();
                    current.right.remove(current.value, current);
                } else if (parent == null) { // root node
                    if (current.left != null) {
                        current.value = current.left.value;
                        current.right = current.left.right;
                        current.left = current.left.left;
                    } else if (current.right != null) {
                        current.value = current.right.value;
                        current.left = current.right.left;
                        current.right = current.right.right;
                    } else {
                        // Single-node tree, do nothing
                    }
                } else if (parent.left == current) {
                    parent.left = (current.left != null) ? current.left : current.right;
                } else if (parent.right == current) {
                    parent.right = (current.left != null) ? current.left : current.right;
                }
                break;
            }
        }
        return this;
    }

    public int getMinValue() {
        BST current = this;
        while (current.left != null) {
            current = current.left;
        }
        return current.value;
    }

    public static void main(String[] args) {
        BST bst = new BST(10);
        bst.insert(5).insert(4).insert(6).insert(15).insert(12).insert(17);

        bst.remove(10, null);
        System.out.println(bst.contains(10)); // false
    }
}
